keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md

```diff
@@ -0,0 +1,832 @@
+
+1. Add a field into APIResource (not mutable) so it could be observed easily(pkg/registry):
+
+```
+type APIResource struct {
+……
+`NamespaceDeletionOrder int64 `json:"namespaceDeletionOrder" protobuf:"varint,11,opt,name=namespaceDeletionOrder"``
+}
+```
+
+In this case, the `NamespaceDeletionOrder` will be associated with the resources naturally and the resource which does not have an opinion will default to a value.
+
+- Pros:
+  - Having it configurable
+  - Consistent between CRD and native type when add support into CRD
+- Cons:
+  - Hard to oversee the overall prioritization across resources
+
+2. Maintain a hard-coded map of DeletionOrder per resources for the resources which have a deletion order preference. Any resources which have no preference opinion would be in the Default category.
+```
+var resourceDeletionOrderPriorities = map[ResourceType]DeletionOrderPriority{
+	"/pods": DeletionOrdeWorkload,
+	"apps.k8s.io/deployments":  DeletionOrdeWorkloadController.
+	"networking.k8s.io/networkpolicies”: DeletionOrdePolicy,
+	……
+})
+
+```
+
+- Pros:
+  - The feature would be fully under control
+  - Have one single place to manage the DeletionOrderPriority across all resources group
+  - Easy for future maintenance
+- Cons:
+  - Not configurable
+  - Not encapsulated (behavior of a type/group defined far away from that type’s “main” code)
+  - Only apply to native types, need separate support for CRD
+
+
+
+During namespace deletion, the namespace collector will traverse resources in the namespace and delete them in ascending priority order based on the `NamespaceDeletionOrder` defined.
+
+### Handling Cyclic Dependencies
+
+Cyclic dependencies can occur if resources within the namespace have finalizers set which conflicts with the DeletionOrderPriority. 
+For example, consider the following scenario:
+
+- Resource A has a finalizer that depends on the deletion of Resource B.
+
+- Resource A is in the earlier DeletionOrderPriority than Resource B.
+
+In this case, the finalizers set would conflict with the `NamespaceDeletionOrder` set for resources and could lead to cyclic dependencies and cause namespace deletion process hanging.
+
+To address this, the system will:
+
+- Attempt to honor the DeletionOrderPriority for resource deletion.
+
+- Monitor the deletion process for each `NamespaceDeletionOrder`. If the process hangs beyond a predefined timeout (e.g., 5 minutes), 
+  it will detect the stall and trigger the deletion attempt for the next `NamespaceDeletionOrder` group.
+
+- If the resources from the next `NamespaceDeletionOrder` group are deleted successfully, retry with the ordered namespace deletion mechanism. If the process hangs again, wait for a predefined timeout and retry.
+
+- If the retry fails over a certain number or the process hangs for too long(an overall timeout), it falls into the fallback mechanism.
```

Comment:
Do we really need an overall timeout?  If we have 4 bands and the very first one gets stuck, you have at worst 20 minutes before you are in the exact same situation as today.